core/chains/cosmos/config.go

```diff
@@ -22,6 +22,10 @@ import (
 
 type CosmosConfigs []*CosmosConfig
 
+func (cs CosmosConfigs) DetectDuplicates(fs CosmosConfigs) (err error) {
+	return fs.ValidateConfig()
```

Comment:
Setting this up as a method even though it doesn't use a receiver seems like a bit of an anti-pattern to me, but it makes the `DetectDuplicates` call in core/services/chainlink/config.go more readable and preserves the class-like association with `CosmosConfigs` and the inheritance-like relationship with `Config`.

Reply:
Hmmmm this pattern is kind of difficult to read. Don't we ultimately only need to pre-validate each config in isolation?

On the other hand, if we are going to have multiple in context, what if we did these dupe checks as part of `SetFrom`, and just returned errors from there? That would be more direct, and we can be precise about only checking the relevant key fields (without picking up e.g. URL validation prematurely).

Reply:
Do you mean something like

```
func (cs *EVMConfigs) SetFrom(fs *EVMConfigs) (err error) {
        // new stuff
        if err = fs.ValidateConfig(); err != nil {
            return
        }
        // end new stuff
	for _, f := range *fs {
		if f.ChainID == nil {
			*cs = append(*cs, f)
		} else if i := slices.IndexFunc(*cs, func(c *EVMConfig) bool {
			return c.ChainID != nil && c.ChainID.Cmp(f.ChainID) == 0
		}); i == -1 {
			*cs = append(*cs, f)
		} else {
			(*cs)[i].SetFrom(f)
		}
	}
}
```

and strip out `DetectDuplicates`?

Reply:
I mean don't use `ValidateConfig()` at all. Just check the key fields directly as we iterate through them here.
```go
func (cs *EVMConfigs) SetFrom(fs *EVMConfigs) (err error) {
	chainIDs := v2.UniqueStrings{}
	for _, f := range *fs {
		if chainIDs.IsDupe(f.ChainID) {
			return v2.NewErrDuplicate(fmt.Sprintf("%d.ChainID", i), *c.ChainID)
		}
		if f.ChainID == nil {
			*cs = append(*cs, f)
		} else if i := slices.IndexFunc(*cs, func(c *EVMConfig) bool {
			return c.ChainID != nil && c.ChainID.Cmp(f.ChainID) == 0
		}); i == -1 {
			*cs = append(*cs, f)
		} else {
			(*cs)[i].SetFrom(f)
		}
	}
}
```
Or, if we want to preserve the property that we get multiple errors at once instead of short circuiting, something like:
```go
func () validateKeys() (err error) {
	// Unique chain IDs
	chainIDs := v2.UniqueStrings{}
	for i, c := range cs {
		if chainIDs.IsDupe(c.ChainID) {
			err = multierr.Append(err, v2.NewErrDuplicate(fmt.Sprintf("%d.ChainID", i), *c.ChainID))
		}
	}

	// Unique node names
	names := v2.UniqueStrings{}
	for i, c := range cs {
		for j, n := range c.Nodes {
			if names.IsDupe(n.Name) {
				err = multierr.Append(err, v2.NewErrDuplicate(fmt.Sprintf("%d.Nodes.%d.Name", i, j), *n.Name))
			}
		}
	}
	return
}
```
To be called as a helper from both `SetFrom` and `ValidateConfig`.

Reply:
There is also the EVM `KeySpecific` config with the same issue:
https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/config/v2/defaults.go#L163-L169